### ISSUE TITLE
Behaviour for receiving a surplus REPONSE

### ISSUE THREAD

**Issue by yschimke** [ReactiveSocket/reactivesocket#116]
If a stream/subscription receives two RESPONSE frames, when only request(1) was requested.  What is the expected behaviour?

Currently I am planning to close the subscription with an error "surplus response" but as I don't want to tear down the whole reactivesocket, I can only cancel the stream.  Tearing down the reactivesocket is also valid, but seems counter to "All other received frames that are not accounted for in previous sections MUST be ignored."

Is this correct?

**Comment by benjchristensen** [comment]
It should `onError` that single stream if flow control is not respected and `cancel` upwards.

@tmontgomery Should this be defined in the spec for how this should be handled?

**Comment by yschimke** [comment]
Great! https://github.com/ReactiveSocket/reactivesocket-cpp/pull/138

**Comment by NiteshKant** [comment]
As a side-note: This is also true for request-channel interaction model on either side (request/response)

I am not sure though whether a ReactiveSocket implementation SHOULD implement this check. At least in JVM land, a receiver of such a response will essentially throw an error which naturally cancels the subscription.

**Comment by yschimke** [comment]
Doesn't that mean the implementation could allow ignoring flow control, and surprising behaviour like surplus messages being buffered by clients or servers leading to memory growth?

**Comment by NiteshKant** [comment]
@yschimke yes, it would result in more data being received by a peer but that isn't really avoidable if we assume rogue peers, eg: Even if a stream is cancelled, one may send more frames. Whether this results in memory growth, not sure, as the implementation has to read invalid frames like this and ignore it anyway.

The downside of forcing this on the implementation means that now the implementation MUST peek and manage flow control per stream instead of delegating it to the consumer. This becomes an overhead if the transport does not care about flow control. Today in java impl we mediate between transport and application flow control, but it isn't necessary in all cases.

So, my suggestion here is to keep it flexible in the protocol spec and give a guideline on how it _may_ be handled. In practice, most consumers will raise error for such an emission and we will achieve exactly the same semantics of cancelling on overflow.

**Comment by benjchristensen** [comment]
Should the required/suggested behavior be documented as per https://github.com/ReactiveSocket/reactivesocket/issues/133 ?
